Work item children

baijum opened the pull request:
fixes https://github.com/fabric8io/fabric8-planner/issues/980

aslakknutsen commented on workitem/workitem_repository.go:
```diff
@@ -314,6 +317,37 @@ func (r *GormWorkItemRepository) List(ctx context.Context, criteria criteria.Exp
+func (r *GormWorkItemRepository) ListChildren(ctx context.Context, parent string) ([]*app.WorkItem, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "children", "query"}, time.Now())
+
+	where := "id in (select work_item_links.target_id from work_item_links left join work_item_link_types on work_item_links.link_type_id = work_item_link_types.id where work_item_link_types.forward_name='relates to' and work_item_links.source_id = ?)"
```
format this a bit?

kwk replied:
@aslakknutsen is right.

```sql
id in (
    SELECT work_item_links.target_id
    FROM work_item_links
    LEFT JOIN work_item_link_types ON work_item_links.link_type_id = work_item_link_types.id
    WHERE
        work_item_link_types.forward_name='relates to'
        AND work_item_links.source_id = ?)
```

When I formatted this I noticed a few things.

1. You should replace the hard coded table names with something like `workitem.WorkItemType{}.TableName()`
1. Instead of doing a `LEFT JOIN` in SQL I would query for the correct link type once and then just filter by ID:
    ```go
    fmt.Sprintf("id in (SELECT id FROM %[1]s WHERE source_id=? AND type_id=?)", link.WorkItemLinks{}.TableName())
    ```
    This should be much much faster than the `LEFT JOIN`.

baijum replied:
Hard coded table name has changed in f44ff7b2fbf2c30c19b945b254c5e1673f83584a and f120c8eae601bc25a32b892ac80128bbc9475338 . Removing left join completed in f44ff7b2fbf2c30c19b945b254c5e1673f83584a

aslakknutsen commented on design/work_item_children.go:
```diff
@@ -0,0 +1,23 @@
+	a "github.com/goadesign/goa/design/apidsl"
+)
+
+var _ = a.Resource("work-item-children", func() {
```
Wonder if this makes more sense in 'work_item_link.go', @kwk ?

kwk replied:
@aslakknutsen I've put the [`list-source-link-types` action](https://github.com/almighty/almighty-core/blob/478f50127c81bea88da6523f203108ba2a042d50/design/resources.go#L58) in the `workitemtype` resource. It returns a list of WITs. The repository functions that are called by the WIT controller on the other hand come from the work item link repositories.

That's why I would create a `list-children` action in the work item resource.

baijum replied:
I have moved the implementation to `workitem/link` package ( Commit: f44ff7b2fbf2c30c19b945b254c5e1673f83584a ) But kept the design in the same file, is that Ok?

kwk replied:
But @aslakknutsen and I recommended to put the list-children action in the work item resource. You've placed it in the `WorkItemLinks` resource. Why?

baijum replied:
I need some clarity here. Is this question about the resource definition in the design package, if so I can move it from `work_item_children.go` to `work_item_link.go`. Please confirm.

kwk replied:
@baijum sorry for the confusion. I hope the BJNs call did clear things up. I will quickly make comments in the PR as well to reflect my suggestions.

kwk commented on workitem/link/link_repository.go:
```diff
@@ -279,3 +283,36 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, lt app.WorkItemLi
+func (r *GormWorkItemLinkRepository) ListWorkItemChildren(ctx context.Context, parent string) ([]*app.WorkItem, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "children", "query"}, time.Now())
+	wilt := WorkItemLinkType{}
+	r.db.Where("forward_name = ?", "relates to").First(&wilt)
```
Why is `"relates to"` chosen to be the link type for children?

baijum replied:
For now, that is hard-coded as per the story ( https://github.com/fabric8io/fabric8-planner/issues/884 ).

kwk replied:
@baijum yes, it may be hard coded but the story doesn't say anything about "relates to" as being the WILT to use. I suggest you create a new WILT to be used explicitly for this purpose because that will allow for a better db migration in the future.

baijum replied:
@kwk Ok, I am going to create a new WILT with the forward name as `parent of` and reverse name as `child of` under `system` category.

baijum commented:
[test]

kwk commented on migration/migration.go:
```diff
@@ -349,6 +349,9 @@ func BootstrapWorkItemLinking(ctx context.Context, linkCatRepo *link.GormWorkIte
 	if err := createOrUpdateWorkItemLinkType(ctx, linkCatRepo, linkTypeRepo, link.SystemWorkItemLinkPlannerItemRelated, "One planner item or a subtype of it relates to another one.", link.TopologyNetwork, "relates to", "is related to", workitem.SystemPlannerItem, workitem.SystemPlannerItem, link.SystemWorkItemLinkCategorySystem); err != nil {
 		return errs.WithStack(err)
 	}
+	if err := createOrUpdateWorkItemLinkType(ctx, linkCatRepo, linkTypeRepo, link.SystemWorkItemLinkTypeParentChild, "One planner item or a subtype of it parent of another one.", link.TopologyNetwork, "parent of", "child of", workitem.SystemPlannerItem, workitem.SystemPlannerItem, link.SystemWorkItemLinkCategorySystem); err != nil {
```
We probably want to also use the link topology "tree" (see [here](https://www.visualstudio.com/en-us/docs/work/reference/link-type-element-reference#link-type-topologies-and-restrictions]). It is currently not allowed by the design folder but we certainly need it sooner than later.

baijum replied:
@kwk I just changed the topology to `tree`. I have updated the enum in the design also. Anything else to do?

kwk commented on workitem/link/link_repository.go:
```diff
@@ -288,7 +288,7 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, lt app.WorkItemLi
 	defer goa.MeasureSince([]string{"goa", "db", "workitem", "children", "query"}, time.Now())
 	wilt := WorkItemLinkType{}
-	r.db.Where("forward_name = ?", "relates to").First(&wilt)
+	r.db.Where("forward_name = ?", "parent of").First(&wilt)
```
I think we should get rid of this whole searching for a built-in link type and instead have the standard link types generated with fixed UUIDs. But that would be a mid-term goal and not relevant for this PR. @xcoulon what do you think?

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -46,195 +45,153 @@ type workItemChildSuite struct {
-
-	// Make sure the database is populated with the correct types (e.g. bug etc.)
-	if err := models.Transactional(DB, func(tx *gorm.DB) error {
-		return migration.PopulateCommonTypes(context.Background(), tx, workitem.NewWorkItemTypeRepository(tx))
```
You still have to populate the types in your code. At least for now.

baijum replied:
@kwk Hmm.. then how the tests are passing now? it use data populated by other tests?

kwk replied:
Yes, probably.

kwk replied:
Fixed in 9e1169f95f2f414dc6bb10aa5da7352610af5249

kwk commented on workitem/link/link_repository.go:
```diff
@@ -307,3 +311,36 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, lt app.WorkItemLi
+	wilt := WorkItemLinkType{}
+	r.db.Where("forward_name = ?", "parent of").First(&wilt)
+
+	where := "id in (select target_id from work_item_links where source_id = ? and link_type_id = ?)"
```
Use `%s` and replace with `workitem.WorkItemLink{}.TableName()` but don't use `work_item_links` directly.

baijum replied:
Fixed in f120c8eae601bc25a32b892ac80128bbc9475338

kwk commented on workitem/link/link_repository.go:
```diff
@@ -307,3 +312,36 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, lt app.WorkItemLi
+func (r *GormWorkItemLinkRepository) ListWorkItemChildren(ctx context.Context, parent string) ([]*app.WorkItem, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "children", "query"}, time.Now())
+	wilt := WorkItemLinkType{}
+	r.db.Where("forward_name = ?", "parent of").First(&wilt)
```
Removed these lines:

```go
	wilt := WorkItemLinkType{}
 	r.db.Where("forward_name = ?", "parent of").First(&wilt)
```

baijum replied:
Fixed in 9cd7414cc72be3ed178ba5e435e78145175b3ffa

kwk commented on workitem/link/link_repository.go:
```diff
@@ -307,3 +312,36 @@ func (r *GormWorkItemLinkRepository) Save(ctx context.Context, lt app.WorkItemLi
+	r.db.Where("forward_name = ?", "parent of").First(&wilt)
+
+	where := fmt.Sprintf("id in (select target_id from %s where source_id = ? and link_type_id = ?)", WorkItemLink{}.TableName())
+	db := r.db.Model(&workitem.WorkItem{}).Where(where, parent, wilt.ID)
```
Replace with

```go
where := fmt.Sprintf(`
id IN (
  SELECT target_id FROM %s
  WHERE source_id = ? AND link_type_id IN (
    SELECT id FROM %s WHERE forward_name = 'parent_of'
  )
)`, WorkItemLink{}.TableName(),WorkItemLinkTypes{}.TableName())
db := r.db.Model(&workitem.WorkItem{}).Where(where, parent)
```

baijum replied:
Fixed in 9cd7414cc72be3ed178ba5e435e78145175b3ffa

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -0,0 +1,263 @@
+	s.userLinkCategoryID = *workItemLinkCategory.Data.ID
+	s.T().Logf("Created link category with ID: %s\n", *workItemLinkCategory.Data.ID)
+
+	s.DB.Unscoped().Delete(&link.WorkItemLinkType{})
```
When you fetch all link types (as suggested) with the forward name `"parent of"` you can remove this list.

baijum replied:
Fixed in 9cd7414cc72be3ed178ba5e435e78145175b3ffa

kwk replied:
@baijum @aslakknutsen for a second I though I might have missed this delete. But we had discussed it and found a way around it.

kwk commented on design/work_item_children.go:
```diff
@@ -0,0 +1,23 @@
+var _ = a.Resource("work-item-children", func() {
+	a.Parent("workitem")
+
+	a.Action("list", func() {
```
Please move the list action to `design/workitems.go` like so, because we don't need another controller just for listing the children.

```go
	a.Action("show", func() {
		a.Routing(
			a.GET("/:id/children"),
		)
		a.Description("List children associated with the given work item.")
		a.Params(func() {
			a.Param("id", d.String, "id")
		})
		a.Response(d.OK, func() {
			a.Media(workItemList)
		})
		a.Response(d.BadRequest, JSONAPIErrors)
		a.Response(d.InternalServerError, JSONAPIErrors)
		a.Response(d.NotFound, JSONAPIErrors)
	})
```

baijum replied:
Fixed in 3c3f20f696492d91f9e39dca8f55cbc2bbe004bc

aslakknutsen commented:
@kwk Do we know anything about how to identify the 'hierarchy' type based on the dynamic space template etc?

kwk commented:
@aslakknutsen no we don't know that yet.

baijum commented:
@kwk I changed code to use `list-children` in the design. Also moved code to `workitems.go`

I am not sure where Goa is adding that `Workitem` suffix.

Let me if any this looks good now. Thanks for the review!

baijum commented:
[test]

baijum commented:
[test]

baijum commented:
[test]

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -0,0 +1,248 @@
+		require.Nil(t, err)
+	}
+	resource.Require(t, resource.Database)
+	suite.Run(t, &workItemChildSuite{DBTestSuite: gormtestsupport.NewDBTestSuite(pwd + "/../config.yaml")})
```
Why do we need `pwd` in here? I think no other test does it.

baijum replied:
I just copied it from `area_test.go`

kwk replied:
I think you can remove it.

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -0,0 +1,248 @@
+	s.userSpaceID = *space.Data.ID
+	s.T().Logf("Created link space with ID: %s\n", *space.Data.ID)
+
+	// Create 3 work items (bug1, bug2, and feature1)
```
The comment should say `// Create 3 work items (bug1, bug2, and bug3)`

baijum replied:
Fixed in 0b3f8d9dbb4b424ba8962bb1f953016e6712f74e

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -0,0 +1,248 @@
+	_, workItemLink := test.CreateWorkItemLinkCreated(s.T(), s.svc.Context, s.svc, s.workItemLinkCtrl, createPayload)
+	require.NotNil(s.T(), workItemLink)
+
+	createPayload2 := CreateWorkItemLink(s.bug1ID, s.bug3ID, s.bugBlockerLinkTypeID)
```
You might as well move the creation of the two links into `SetupTest`. Then you only need to store the parent bug ID (`bug1ID`) in the suite, because you're comparing the bug's titles and not their IDs below. Which means, that you can shrink these variables:

```go
	bug1ID               uint64
 	bug2ID               uint64
 	bug3ID               uint64
 	userLinkCategoryID   uuid.UUID
 	bugBlockerLinkTypeID uuid.UUID
 	userSpaceID          uuid.UUID
```

to just 

```go
	bug1ID               uint64
```

baijum replied:
Fixed in 0b3f8d9dbb4b424ba8962bb1f953016e6712f74e

kwk commented on controller/workitem.go:
```diff
@@ -493,8 +493,46 @@ func ConvertWorkItem(request *goa.RequestData, wi *app.WorkItem, additional ...W
+
+// WorkItemIncludeChildren adds relationship about children to workitem (include totalCount)
+func WorkItemIncludeChildren(request *goa.RequestData, wi *app.WorkItem, wi2 *app.WorkItem2) {
+	wi2.Relationships.Children = CreateChildrenRelation(request, wi)
```
Why not simply do the following instead of calling three functions? Are those functions used somewhere else? They are exported so I'm guessing you had plans for them, no?

```go
childrenRelated := rest.AbsoluteURL(request, app.WorkitemHref(wi.ID)) + "/children"
wi2.Relationships.Children = &app.RelationGeneric{
  Links: &app.GenericLinks{
    Related: &childrenRelated,
  },
}
```

baijum replied:
Fixed in ab2a250fdfe8967e752a684daa7c51d0f793538c

baijum replied:
Fixed in c4f8ee2163a60097de8e97a5a6f1b8b0e5f12f3e

kwk commented on migration/migration.go:
```diff
@@ -422,6 +422,9 @@ func BootstrapWorkItemLinking(ctx context.Context, linkCatRepo *link.GormWorkIte
 	if err := createOrUpdateWorkItemLinkType(ctx, linkCatRepo, linkTypeRepo, spaceRepo, link.SystemWorkItemLinkPlannerItemRelated, "One planner item or a subtype of it relates to another one.", link.TopologyNetwork, "relates to", "is related to", workitem.SystemPlannerItem, workitem.SystemPlannerItem, link.SystemWorkItemLinkCategorySystem, space.SystemSpace); err != nil {
 		return errs.WithStack(err)
 	}
+	if err := createOrUpdateWorkItemLinkType(ctx, linkCatRepo, linkTypeRepo, spaceRepo, link.SystemWorkItemLinkTypeParentChild, "One planner item or a subtype of it parent of another one.", link.TopologyTree, "parent of", "child of", workitem.SystemPlannerItem, workitem.SystemPlannerItem, link.SystemWorkItemLinkCategorySystem, space.SystemSpace); err != nil {
```
I think this comment is a bit odd: `One planner item or a subtype of it parent of another one.`

baijum replied:
Fixed in c4f8ee2163a60097de8e97a5a6f1b8b0e5f12f3e

kwk replied:
Sorry, but the sub-type portion is actually important: `One planner item which is a parent of another one.` needs to be `One planner item or a sub-type of it which is a the parent of another one.`

baijum replied:
Fixed in f3adda9e3adcf3b1aee985fd375762cf8fb6a954

kwk replied:
Fixed in f3adda9

kwk commented on controller/work-item-children-blackbox_test.go:
```diff
@@ -45,12 +45,14 @@ type workItemChildSuite struct {
-	bugBlockerLinkTypeID uuid.UUID
-	userSpaceID          uuid.UUID
+	bug1ID uint64
+	/*
```
Please remove this comment

baijum replied:
Fixed in e8a53e6aff98aef44d338943dd93c768514a96fd

aslakknutsen commented:
@baijum Could you create a follow up Issue to redo the this to use some form of link type when we get there?

baijum commented:
> @baijum Could you create a follow up Issue to redo the this to use some form of link type when we get there?

Can you please give more details about this?

aslakknutsen commented:
@baijum The space template defines the forward and reverse names and any link per say. Children can't really be based on that someone chose to call their link parent-of. We most likely need some form of link category or similar to state what this 'random link type defined by the space template' is suppose to represent
